### ISSUE TITLE
Frequent SQLite operations when loading analysis data lead to unnecessary amount of syscalls

### ISSUE THREAD

**Issue by onyxmaster** [dotnet/roslyn#22339]
**Version Used**: 
Released with VS2017 15.3.5, also custom-built from dev15.5.x

**Description**:

Consider a moderately large solution (our solution has ~160 projects, Roslyn one will do as an example).
I'm no expert in Roslyn internals, so this is more of a cargo cult performance research, but please bear with me.
Some time after solution is loaded, the `DiagnosticIncrementalAnalyzer` attempts to load initial data using `ProjectState.LoadInitialAnalysisDataAsync()`, which in its turn attempts to load multiple `DiagnosticAnalysisResult`s from the analysis database. This (in VS) invokes `SQLitePersistentStorage.ReadStreamAsync`, that invokes SQLite. By default, SQLite uses [journal mode](https://sqlite.org/pragma.html#pragma_journal_mode) *DELETE*, which means that journal is deleted when not used. When [preparing a transaction](https://github.com/mackyle/sqlite/blob/5e9d7dc9750e4bf98e7709fa09628cd156d72b80/src/btree.c#L3259), SQLite [acquires a shared lock for a btree](https://github.com/mackyle/sqlite/blob/5e9d7dc9750e4bf98e7709fa09628cd156d72b80/src/btree.c#L2951), and, consequentially, [a pager](https://github.com/mackyle/sqlite/blob/f4d982b49fb4ac64e708f2552ce1c32ad82c5472/src/pager.c#L5140), where a [series](https://github.com/mackyle/sqlite/blob/f4d982b49fb4ac64e708f2552ce1c32ad82c5472/src/pager.c#L5036) of [checks](https://github.com/mackyle/sqlite/blob/f4d982b49fb4ac64e708f2552ce1c32ad82c5472/src/pager.c#L3334) is done, testing if database needs recovery (either has rollback journal or uses WAL). Unfortunately, since the journal is deleted after each successful transaction, even read-only transactions perform an extra 2 system calls (there are actually more than 2 extra calls in non-readonly case, but these ones would suffice for now) -- the `FileIoCreate` for files named "storage.ide-journal" and "storage.ide-wal".

Since there is a lot of transactions, over the course of a 10-second ETW trace captured after solution finished loading and Roslyn kicked in to load diagnostics, about 200000 of each `FileIoCreate` syscalls were recorded, making it about 200000 * 2 / 10 = 40K syscalls/s.
On all machines I have access to, this syscall is not fast enough, and disk speed (RAMDrive made no difference vs PCI-E SSD), installed driver-level filters (Defender realtime protection was turned off) or filesystem (FAT32 made no difference vs NTFS) do not matter much. Over the same trace, out of ~9600ms CPU time spent by `DiagnosticsDataSerializer.DeserializeAsync` and its descendants, the combination of `kernel32!GetFileAttributesExW` and `kernel32!GetFileSize` account for ~4200ms of CPU time, which is roughly half of the time spent deserializing data.

**Proposal**:
While I believe that some kind of read combining for inital analysis data loading would be a better solution, I consider this a larger effort that should be addressed later, so my proposal is a bit simpler: enable [write-ahead logging](https://sqlite.org/wal.html). I did a quick search over issues/PRs and didn't find any explicit mentions of WAL, except #21800, which mentions WAL by itself, but doesn't introduce it. I did a simple test, and enabling WAL reduced the initial analysis load time by about 40%.

**Final notes**:
As I'm no expert on neither Roslyn and its inner workings, nor SQLite, it may be so that my analysis is incorrect. Also, there might be reasons (unknown to me) that make WAL not an option. If this is the case, please excuse me for spending your time on reading this.

**Technical details**:
Performance traces were recorded with PerfView with in-memory circular buffer to prevent event loss due to lots of async transfer events. I did not record thread time, because I wasn't researching stalls/locks.
***CPU analysis***:
PerfView `CPU Stacks` settings:
```
GroupPats: [group CLR/OS entries]
FoldPats: mscorlib!;mscorlib.ni!;sqlitepclraw.core!;sqlitepclraw.provider.e_sqlite3!
```

Before:
```
Name                                                                                                                                   	Inc %	     Inc
 module microsoft.codeanalysis.workspaces.desktop <<microsoft.codeanalysis.workspaces.desktop!SQLitePersistentStorage.ReadStreamAsync>>	 86,7	   9 168
+ module e_sqlite3 <<e_sqlite3!?>>                                                                                                     	 72,4	   7 661
|+ OS <<kernelbase!GetFileAttributesExW>>                                                                                              	 42,2	   4 463
|+ OS <<kernelbase!GetFileSize>>                                                                                                       	  6,2	     658
|+ OS <<kernelbase!LockFileEx>>                                                                                                        	  4,2	     448
|+ OS <<kernelbase!ReadFile>>                                                                                                          	  3,6	     383
|+ OS <<kernelbase!UnlockFileEx>>                                                                                                      	  2,7	     281
```

After:
```
Name                                                                                                                                   	Inc %	     Inc
 module microsoft.codeanalysis.workspaces.desktop <<microsoft.codeanalysis.workspaces.desktop!SQLitePersistentStorage.ReadStreamAsync>>	 75,8	   8 151
+ module e_sqlite3 <<e_sqlite3!?>>                                                                                                     	 40,9	   4 398
|+ OS <<kernelbase!LockFileEx>>                                                                                                        	  8,7	     940
|+ OS <<kernelbase!UnlockFileEx>>                                                                                                      	  5,4	     585
```

***Syscall analysis***:
PerfView `Files Stacks` settings:
```
GroupPats: [group module entries]
FoldPats: wow64!;wow64cpu!;ntdll!;ntoskrnl!;kernelbase!;fltmgr!;fileinfo!;sqlitepclraw.provider.e_sqlite3!;sqlitepclraw.core!;mscorlib!
```

```
Name                                                                                                                                   	Inc %	       Inc
 module microsoft.codeanalysis.workspaces.desktop <<microsoft.codeanalysis.workspaces.desktop!SQLitePersistentStorage.ReadStreamAsync>>	 97,0	   404 878
+ module e_sqlite3 <<e_sqlite3!?>>                                                                                                     	 97,0	   404 878
 + FileOpenOrCreate ...\v15\sqlite3\storage.ide-journal                              	 48,5	   202 439
 + FileOpenOrCreate ...\v15\sqlite3\storage.ide-wal                                  	 48,5	   202 439
```

**Comment by Pilchie** [comment]
Tagging @heejaechang and @sharwell 

Also @CyrusNajmabadi if he wants to take a look at this and the PR.

**Comment by onyxmaster** [comment]
The original proposal is preserved for historical reasons, now thanks to @sharwell I believe that WAL is not needed if locking mode can be changed, the discussion is at #22340.

**Comment by onyxmaster** [comment]
I believe this can be closed since #22340 is merged.
